blues/sparrow-reference-web-app
# feat: sensor details page mock api hookup

# Problem Context

AC:

* Sensor Summary page should retrieve all environment variables and other pertinent device-specific data for the sensor and display it within the Sensor Detail page
* Page should also be able to issue POSTs/PUTs for saving environment variable data

## Changes

Hooked up both the main dashboard and sensor details pages to our API endpoints and Notehub (both mocked and live data). There's now data transformation files for both latest sensor data (for the dashboard snapshots) and historical sensor data (for the charts on the individual sensors).
New models for Notehub data shapes.
New constants to handle when data's missing in Notehub or to fill in empty spaces in the UI.
Rejiggered the `VoltageChart` to accept more configs from the component passing data to it to make it more reusable.
Added "rules" to the form so when a user does not give a new sensor name and location, form submission will fail and display error messages.

## Screenshot (if applicable)

![Screen Shot 2022-01-06 at 6 09 14 PM](https://user-images.githubusercontent.com/20400845/148465415-29e6ed14-594a-4464-a9ee-ee53dccb955a.png)

## Testing

Unit / integration / e2e tests?

Holding off on unit tests until after a major refactor of code inside of the `/lib/` (soon to be services) folder is completed.

Steps to test manually?

1. Make sure all local env vars are set in `.env.local`
2. Run `yarn dev` and visit localhost:4000/ in the browser
3. Observe all the sensor cards now have different data (this is actual data from Notehub)
4. Click on the card named TEST_NAME
5. Wait for the charts to load in TEST_NAME
6. See all the charts and data loaded (this is real Notehub data again)
7. Click into the next tab and try changing the name 
8. Refresh the page to see the updated sensor name or go to the dashboard and see the updated name (I tried a bunch of ways to get this to work without a hard refresh on the component to no avail and eventually got frustrated and threw in the towel)
9. Change the name back to TEST_NAME on the sensor for the next dev testing this

Any other sorts of testing notes to include?

Currently the last viable sparrow event data from Ray's Sparrows dates back to Nov 30, which is why I'm pulling historical data from 40 days ago to populate just one sensor details page. Hopefully in the future, we'll be able to make that default for historical data configurable to the last seven days or whatever the user chooses. That should make the process faster to load pages as well.

## Any other related PRs 

n/a


## Ticket(s)

https://www.pivotaltracker.com/story/show/180562645
